question22/main.js

```diff
@@ -0,0 +1,188 @@
+    userDataColumn[key].sort && th.appendChild(createSortButton());
+    fragmentThElement.appendChild(tr).appendChild(th); 
+  });
+
+  thead.appendChild(fragmentThElement);
+  return thead;
+}
+
+const getTableRowFragment = (userData) => {
+  const tableRowfragment = document.createDocumentFragment();
+  for (let i = 0; i < userData.length; i++) {
+    const tr = createElementWithClassName("tr", "tr");
+    tableRowfragment.appendChild(tr).appendChild(getTableDataFragment(userData[i]));
+  }
+  return tableRowfragment;
+}
+
+const getTableDataFragment = (userData) => {
+  const tableDatafragment = document.createDocumentFragment();
+  Object.keys(userDataColumn).forEach((key) => {
+    const td = createElementWithClassName("td", "td");
+    td.textContent = userData[key];
+    tableDatafragment.appendChild(td);
+  });
+  return tableDatafragment;
+};
+
+const createSortButton = () => {
+  const sortButton = createElementWithClassName("button", "sortButton");
+  sortButton.dataset.status = "default";
+  sortButton.id = "js-sortbtn";
+  return sortButton;
+}
+
+const sortButtonClickEvent = () => {
+  const tbody = document.querySelector('tbody');
+  const sortButtons = [...document.querySelectorAll(".sortButton")];
+  const trElement = [...document.querySelectorAll("tbody > tr")];
+  sortButtons.forEach((sortButton)=> {
+    sortButton.addEventListener('click', (e) => {
+      resetSortbuttonStatus(sortButtons, e.target);
+      const clickedCellIndex = e.target.parentElement.cellIndex;
+      const nextStatus = switchSortStatus(e.target);
+      sortButton.dataset.status = nextStatus;
+      const sortedRows = sortFunc(e.target,clickedCellIndex,trElement);
+      while (tbody.firstChild) {
+        tbody.removeChild(tbody.firstChild);
+    }
+      sortedRows.forEach((row) => {
+        tbody.appendChild(row);
+      });
+    });
+  })
+}
+
+const resetSortbuttonStatus = (sortButtons, target) => {
+  sortButtons.forEach((button)=> {
+    if (button !== target) {
+      button.dataset.status = Sort.Default;
+    }
+  })
+}
+
+const Sort = {
+  Default: "default",
+  Asc: "asc",
+  Desc: "desc"
+};
+
+const switchSortStatus = (target) => {
+  switch (target.dataset.status) {
+    case Sort.Default:
+      return Sort.Asc;
+    case Sort.Asc:
+      return Sort.Desc;
+    case Sort.Desc:
+      return Sort.Default; 
+    default:
+      return Sort.defaul;
+  }
+}
+
+const sortFunc = (target,clickedCellIndex,defaultRows) => {
+  const defaultTrElement = [...defaultRows];
+  const currentStatus = target.dataset.status;
+  if (currentStatus === Sort.Default) {
+      return defaultRows;
+  }
+  if (currentStatus === Sort.Asc){
+      return defaultTrElement.sort((a,b) => a.children[clickedCellIndex].textContent - b.children[clickedCellIndex].textContent);
+  } 
+  if (currentStatus === Sort.Desc){
+      return defaultTrElement.sort((a,b) => b.children[clickedCellIndex].textContent - a.children[clickedCellIndex].textContent);
+  } 
```

Comment:
IMO
It looks like it can be put together into a switch.
I'll leave it up to you to fix this one☺️

```js
 switch (currentStatus) {
    case Sort.Asc:
      return defaultTrElement.sort(
        (a, b) =>
          a.children[clickedCellIndex].textContent -
          b.children[clickedCellIndex].textContent
      );
    case Sort.Desc:
      return defaultTrElement.sort(
        (a, b) =>
          b.children[clickedCellIndex].textContent -
          a.children[clickedCellIndex].textContent
      );

    default:
      return defaultRows;
  }
```

Reply:
Thank you for your ideas! I felt the swich text was more readable, so I fixed it! 
fixed commit: https://github.com/csakyo/homework/pull/37/commits/e5f9c71ce0391482590eedc14a00bfedf20ccfbd

Reply:
I felt that the code here is very good, but it would be better to define it as an object as much as possible because the more arguments it has, the less readable it may become.
[関数の引数が多い場合、Patterns Objectを利用する](https://zenn.dev/chida/articles/ff3e073bce9306#7.%E9%96%A2%E6%95%B0%E3%81%AE%E5%BC%95%E6%95%B0%E3%81%8C%E5%A4%9A%E3%81%84%E5%A0%B4%E5%90%88%E3%80%81patterns-object%E3%82%92%E5%88%A9%E7%94%A8%E3%81%99%E3%82%8B)

```js
const addClickEventToSortButton = () => {
  const sortButtons = [...document.querySelectorAll(".sortButton")];
  const trElement = [...document.querySelectorAll("tbody > tr")];
  sortButtons.forEach((sortButton) => {
    sortButton.addEventListener("click", (e) => {
      clickSortButton({
        target: e.target,
        sortButtons: sortButtons,
        trElement: trElement
      });
    });
  });
};

const clickSortButton = ({ target, sortButtons, trElement }) => {
  const tbody = document.querySelector("tbody");
  resetSortbuttonStatus(sortButtons, target);
  const clickedCellIndex = target.parentElement.cellIndex;
  const nextStatus = switchSortStatus(target);
  target.dataset.status = nextStatus;
  const sortedRows = sortFunc(target, clickedCellIndex, trElement);
  while (tbody.firstChild) {
    tbody.removeChild(tbody.firstChild);
  }
  sortedRows.forEach((row) => {
    tbody.appendChild(row);
  });
};

```
It seems that there is also a way to write with handleEvent.
```js
const addClickEventToSortButton = () => {
  const sortButtons = [...document.querySelectorAll(".sortButton")];
  const trElement = [...document.querySelectorAll("tbody > tr")];
  sortButtons.forEach((sortButton) => {
    sortButton.addEventListener("click", {
      sortButtons: sortButtons,
      trElement: trElement,
      handleEvent: clickSortButton
    });
  });
};

function clickSortButton(event) {
  const tbody = document.querySelector("tbody");
  resetSortbuttonStatus(this.sortButtons, event.target);
  const clickedCellIndex = event.target.parentElement.cellIndex;
  const nextStatus = switchSortStatus(event.target);
  event.target.dataset.status = nextStatus;
  const sortedRows = sortFunc(event.target, clickedCellIndex, this.trElement);
  while (tbody.firstChild) {
    tbody.removeChild(tbody.firstChild);
  }
  sortedRows.forEach((row) => {
    tbody.appendChild(row);
  });
}

```